external-crates/move/crates/move-compiler/src/expansion/aliases.rs

```diff
@@ -205,78 +293,117 @@ impl AliasMapBuilder {
-        result
+        self.add_member_alias_(alias, ident, member, kind, /* is_implicit */ true)
     }
 }
 
+macro_rules! resolve_alias {
+    ($map:expr, .$namespace:ident, $name:expr) => {{
+        let mut current_scope = Some($map);
+        let name = $name;
+        loop {
+            let Some(scope) = current_scope else {
+                break None;
+            };
+            let Some(entry) = scope.$namespace.get(name).copied() else {
+                current_scope = scope.previous.as_mut().map(|x| &mut **x);
+                continue;
+            };
+            // Note, might have already been removed by a different `$namespace` resolution
+            scope.unused.remove(&(*name, entry).into());
+
+            let original_name = scope.$namespace.get_full_key(&name).unwrap();
+            break Some((original_name, entry));
+        }
+    }};
+}
+
 impl AliasMap {
     pub fn new() -> Self {
         Self {
-            aliases: UniqueMap::new(),
+            unused: BTreeSet::new(),
+            leading_access: UniqueMap::new(),
+            module_members: UniqueMap::new(),
             previous: None,
         }
     }
 
-    pub fn get(&mut self, name: &Name) -> Option<AliasEntry> {
-        if let Some((entry, used)) = self.aliases.get_mut(name) {
-            *used = true;
-            // Type parameters are never resolved. We track them to apply appropriate shadowing.
-            if matches!(entry, AliasEntry::TypeParam) {
-                return None;
-            } else {
-                return Some(entry.clone());
-            }
+    pub fn resolve_leading_access(&mut self, name: &Name) -> Option<(Name, LeadingAccessEntry)> {
+        let (name, entry) = resolve_alias!(self, .leading_access, name)?;
+        match &entry {
+            LeadingAccessEntry::Module(_)
+            | LeadingAccessEntry::Address(_)
+            | LeadingAccessEntry::Member(_, _) => Some((name, entry)),
+            // For legacy reasons, don't resolve type parameters
+            LeadingAccessEntry::TypeParam => None,
         }
+    }
 
-        let mut current_scope: Option<&mut Box<AliasMap>> = self.previous.as_mut();
-        while let Some(scope) = current_scope {
-            if let Some((entry, used)) = scope.aliases.get_mut(name) {
-                *used = true;
-                // Type parameters are never resolved. We track them to apply appropriate shadowing.
-                if matches!(entry, AliasEntry::TypeParam) {
-                    return None;
-                } else {
-                    return Some(entry.clone());
-                }
-            }
-            current_scope = scope.previous.as_mut();
+    pub fn resolve_call(&mut self, name: &Name) -> Option<(Name, MemberEntry)> {
+        let (name, entry) = resolve_alias!(self, .module_members, name)?;
+        match &entry {
+            MemberEntry::Member(_, _) => Some((name, entry)),
+            // For legacy reasons, don't resolve type parameters
```

Comment:
Is this comment correct, or do we do it for shadowing?

Reply:
I really have no clue why we don't resolve type parameters lol

I meant for code-legacy reasons. Like the type parameter is here only for shadowing currently. Its a bit wonky